Actionitems

kiranmisner opened the pull request:
IN Progress: New/Added Action items needed to be compared for a duplicate and also added to the request object that is sent as a response

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -165,6 +166,7 @@ export default class AlumniMentorship extends Component {
 
     handleMenuClick = (e, { id }) => this.setState({ activeItem: id })
 
+
```
remove extra line

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -384,14 +386,19 @@ export default class AlumniMentorship extends Component {
  *                  used to disable approval, preventing double booking
  */
 class RequestCards extends Component {
+    
```
remove extra lines

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -577,7 +606,7 @@ class RequestCards extends Component {
 
     async submitFinalNote(e) {
-        let requests = await makeCall({
+        let requests = await  makeCall({
```
nit: remove extra space

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -587,7 +616,7 @@ class RequestCards extends Component {
     handleValueChange(e, {name, value}) {
         e.preventDefault();
-        this.setState({
+        this.setState({ 
```
nit: extra space

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -600,6 +629,22 @@ class RequestCards extends Component {
     }
 
+    getActionItemsInput(selection) {
+        console.log(selection);
```
remove log here

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -600,6 +629,22 @@ class RequestCards extends Component {
+    }
+
+    async submitActionItems() {
+        let requests = await  makeCall({
```
extra space between await and makeCall

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -641,6 +686,42 @@ class RequestCards extends Component {
+                                Please select up to 3 action items for {this.state.requestDetails.student.name}
+                                <PooledMultiSelectDropdown 
+                                 allowAddition='true'
+                                 endpoint={`/drop/actionitems`}
```
nit: let's have actionItems be camelCase

reshadbinharun replied:
This request might still fail because of case, should be `/drop/actionItems`

reshadbinharun commented on client/src/components/PooledMultiSelectDropdown.js:
```diff
@@ -12,6 +12,7 @@ props:
 export default class PooledMultiSelectDropdown extends Component {
     constructor(props){
         super(props)
+        console.log(props); 
```
nit: remove log

reshadbinharun commented on client/src/components/alumni_profile_update_modals/JobTitleUpdateModal.js:
```diff
@@ -15,6 +15,7 @@ import swal from "sweetalert";
 export default class JobTitleUpdateModal extends Component {
     constructor(props){
         super(props)
+        console.log("job props: ", props);
```
nit: remove log

reshadbinharun commented on server/routes/dropdown.js:
```diff
@@ -130,4 +131,21 @@ router.get('/colleges/:country', async (req, res) => {
   }
 })
 
+router.get('/actionitems', async (req, res) => {
```
nit: camelCase here too!

reshadbinharun commented on server/routes/utilMongoose.js:
```diff
@@ -773,4 +773,37 @@ router.get('/data/clear/all', async (req, res, next) => {
+router.post('/addActionItems/', async (req, res, next) => {
+    try {
+        const name = req.body.name;
+        const user_instance = new actionItemSchema(
```
user_instance -> actionItem_instance

reshadbinharun commented on server/routes/utilMongoose.js:
```diff
@@ -773,4 +773,37 @@ router.get('/data/clear/all', async (req, res, next) => {
+        );
+        await user_instance.save();
+
+        var student_instance = new studentSchema(
```
We don't need to create an action item instance here

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -600,6 +624,21 @@ class RequestCards extends Component {
+        let requests = await makeCall({
+            requestId: this.state.requestDetails._id,
+            actionItems: this.state.existingActionItems
+        }, `/request/actionItems/${this.props.userId}/${this.props.timeOffset}`, 'patch')
```
Can we add a `swal` modal for success/failure? examples can be found in any of the *UpdateModals

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -641,6 +680,42 @@ class RequestCards extends Component {
+                                />
+                            </Grid.Column>
+                            <Grid.Column>
+                                Please select up to 3 action items for {this.state.requestDetails.student.name}
```
Let's add this text in `Modal.Description` tag?

reshadbinharun replied:
bump on this comment

reshadbinharun commented on server/models/requestSchema.js:
```diff
@@ -26,6 +26,7 @@ const requestSchema = new Schema(
     studentNote: {type: String},
     alumniNote: {type: String},
     finalNote: {type: String},
+    actionItems: [{type: String}]
```
missing comma here

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,53 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
     }
 })
 
+router.patch('/actionItems/:id/:timeOffset', passport.authenticate('jwt', {session: false}), async (req, res, next) => {
```
We might not need a :timeOffset url param here...

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,53 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+    let requests = []
+    try {
+        let request = await requestSchema.findById(requestId);
+        request.actionItems = actionItems;
```
This might need to be `request.actionsItems.push(actionItems)` since an actionItem is an object, but the field on the schema is an array

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,53 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+        let request = await requestSchema.findById(requestId);
+        request.actionItems = actionItems;
+        await request.save()
+        for (let status of conditions) {
```
We might not need to send back requests in the response and something like `res.status(200).send({'message' : "Added action items"});`  might be sufficient

reshadbinharun replied:
So we can remove the code where we're fetching the all the requests for the alumni

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,53 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+
+});
+
+router.post('/addActionItem/', async (req, res, next) => {
```
I understand the context for the conversation from earlier now! We can remove this end-point, and have any new actionItems be created in the PATCH `/actionItems/:id` request above.
An example of how new actionItems and existing actionItems can be passed in req.body can be found in `alumni.js`'s `/interests/add/:id` PATCH end-point.

kiranmisner commented:
Not sure if I had added it to the pull request but the branch is definitely updated all the way. I made the following changes:
- Action Items added to request onSubmit using generateNewActionItems and getUniqueActionItems. 
- Depending on success, swal alert shown to update user. 
- Deleted code that updated onChange

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -600,6 +624,34 @@ class RequestCards extends Component {
+        if (!requests || requests.error) {
+            swal({
+                title: "Error!",
+                text: "There was an error updating this request, please try again.",
```
Let's have the error/success messages be specific to actionItems

reshadbinharun commented on client/src/components/PooledMultiSelectDropdown.js:
```diff
@@ -72,7 +72,7 @@ export default class PooledMultiSelectDropdown extends Component {
         })
     }
-
+    
```
stray whitespace here, please remove

reshadbinharun replied:
nit: still has stray white spaces!

reshadbinharun replied:
bump on this nit! [this](https://packagecontrol.io/packages/Highlight%20Trailing%20Whitespace) might help

reshadbinharun commented on server/routes/dropdown.js:
```diff
@@ -8,6 +8,7 @@ var companySchema = require('../models/companySchema');
 var majorSchema = require('../models/majorSchema');
 var interestsSchema = require('../models/interestsSchema');
 var COUNTRIES = require("../countries").COUNTRIES
+var actionitemSchema = require("../models/actionItemSchema");
```
let declare this in camelCase as `actionItemSchema`

kiranmisner commented:
<img width="825" alt="Screen Shot 2020-08-10 at 4 16 27 PM" src="https://user-images.githubusercontent.com/59927955/89840394-02795380-db25-11ea-9364-086efb25e48e.png">
<img width="898" alt="Screen Shot 2020-08-10 at 4 16 40 PM" src="https://user-images.githubusercontent.com/59927955/89840400-05744400-db25-11ea-8b4a-51a01c3b879c.png">
<img width="1421" alt="Screen Shot 2020-08-10 at 4 16 48 PM" src="https://user-images.githubusercontent.com/59927955/89840408-086f3480-db25-11ea-88a3-0c7151c82534.png">

reshadbinharun commented on server/routes/dropdown.js:
```diff
@@ -130,4 +131,21 @@ router.get('/colleges/:country', async (req, res) => {
 
+router.get('/actionItems', async (req, res) => {
+  try {
+    let actionitems = await actionitemSchema.find()
```
nit: camelCase for `actionItems`

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,55 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+    if (newActionItems.length) {
+        for (let i = 0; i < newActionItems.length; i++) {
+            // check to see if action items name already exists
+            let actionItemsExists = await interestsSchema.find({name: newInterests[i].value})
```
`actionItemsSchema` not `interestsSchema`

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,55 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+        let requestId = req.body.requestId;
+        let request = await requestSchema.findById(requestId);
+        const existingActionItems = req.body.existingActionItems
+        const newActionItems = []
```
Let's also send newActionItems and parse them out of `req.body`

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,56 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+        const existingActionItems = req.body.existingActionItems
+        const newActionItems = req.body.newActionItems || []
+        let actionItemsToAdd = await generateNewAndExistingActionItems(existingActionItems, newActionItems);
+        console.log("To add: ",actionItemsToAdd);
```
remove log here!

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,55 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+                await newActionItemCreated.save()
+                existingActionItemsRecords.push(newActionItemCreated)
+            } else {
+                // user accidentally added an interest that already exists as a new interest
```
nit: actionItem

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,55 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+const generateNewAndExistingActionItems = async (existingActionItems, newActionItems) => {
+    const existingActionItemsIds = existingActionItems.map(actionItem => actionItem.value).flat()
+    let existingActionItemsRecords = await actionItemSchema.find().where('_id').in(existingActionItemsIds).exec()
+    // create action itmes added
```
nit: items*

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -600,6 +624,35 @@ class RequestCards extends Component {
+            requestId: this.state.requestDetails._id,
+            existingActionItems: this.state.existingActionItems,
+            newActionItems: this.state.newActionItems
+        }, `/request/actionItems/${this.state}/`, 'patch')
```
`this.state` is an object here, and should break the url composition

reshadbinharun replied:
and pass it as a url parameter here like `/${this.state.requestDetails._id}` (this.state here is an object and cannot be parsed correctly)

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -641,6 +694,44 @@ class RequestCards extends Component {
+                            Cancel
+                        </Button>
+                        <Button onClick={this.submitActionItems.bind(this)} primary>
+                            Submit
```
Let's disable submit button if there are no new or existing action items!

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -600,6 +624,35 @@ class RequestCards extends Component {
+
+    async submitActionItems() {
+        let requests = await makeCall({
+            requestId: this.state.requestDetails._id,
```
You can avoid sending requestId in body...

reshadbinharun commented on server/routes/requests.js:
```diff
@@ -322,4 +323,55 @@ router.get('/getConfirmed/:id/:timeOffset', passport.authenticate('jwt', {sessio
+
+router.patch('/actionItems/:id/', passport.authenticate('jwt', {session: false}), async (req, res, next) => {
+    try {
+        let requestId = req.body.requestId;
```
one you pass the requestId as url params, you can get it here like `req.params.id`

reshadbinharun commented on server/.env.tpl:
```diff
@@ -19,4 +19,4 @@ AWS_SECRET_ACCESS_KEY=
 SENDGRID_KEY=
-DEV_MODE=
\ No newline at end of file
+DEV_MODE=true
```
thanks!

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -540,6 +553,19 @@ class RequestCards extends Component {
     }
 
+    toggleActionItemModal(e) {
+        let requestDetails = this.props.requests.find(request => request._id === e.currentTarget.getAttribute('requestid'))
```
We can remove everything from here except
`this.setState({showActionItemsModal: !this.state.showActionItemsModal })`
There is no actionItems state we need to store

reshadbinharun commented on client/src/components/AlumniMentorship.js:
```diff
@@ -641,6 +693,45 @@ class RequestCards extends Component {
+                            Cancel
+                        </Button>
+                        <Button onClick={this.submitActionItems.bind(this)} 
+                         disabled = {this.state.existingActionItems || this.state.newActionItems} primary>
```
So the disabled conditional here should be `!(this.state.existingActionItems && this.state.existingActionItems.length) && !(this.state.newActionItems && this.state.newActionItems.length)` Let me know if that makes sense!
Also nit: let indent on `disabled` and have primary be on a new line
